From ac9c53cbdf6538a11855731311c743fb9b09c1f3 Mon Sep 17 00:00:00 2001
From: Bernhard Rosenkraenzer <Bernhard.Rosenkranzer@linaro.org>
Date: Mon, 12 Dec 2011 13:25:14 +0059
Subject: [PATCH] variablespeed: Fix aliasing violation

Signed-off-by: Bernhard Rosenkraenzer <Bernhard.Rosenkranzer@linaro.org>
Change-Id: Ic62c65b41e83f556ce228c3f724433ebace58168

ex: Fix mismatch between header and implementation

The header defines GetChannelCount() as returning a size_t while
the implementation defines it as returning a SLuint32.

Change-Id: Iaa837d9a39540c338582496a33a4bbd8e3c63b55
Signed-off-by: Bernhard Rosenkraenzer <Bernhard.Rosenkranzer@linaro.org>
---
 variablespeed/jni/variablespeed.cc | 6 ++++--
 variablespeed/jni/variablespeed.h  | 2 +-
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/variablespeed/jni/variablespeed.cc b/variablespeed/jni/variablespeed.cc
index 73ac609..9a15bcf 100644
--- a/variablespeed/jni/variablespeed.cc
+++ b/variablespeed/jni/variablespeed.cc
@@ -242,11 +242,13 @@ static void ReadSampleRateAndChannelCount(CallbackContext *pContext,
         if (value) {
           OpenSL(decoderMetadata, GetValue, i, valueSize, value);
           if (strcmp((char*) keyInfo->data, ANDROID_KEY_PCMFORMAT_SAMPLERATE) == 0) {
-            SLuint32 sampleRate = *(reinterpret_cast<SLuint32*>(value->data));
+            SLuint32 sampleRate;
+            memcpy(&sampleRate, value->data, sizeof(SLuint32));
             LOGD("sample Rate: %d", sampleRate);
             *sampleRateOut = sampleRate;
           } else if (strcmp((char*) keyInfo->data, ANDROID_KEY_PCMFORMAT_NUMCHANNELS) == 0) {
-            SLuint32 channels = *(reinterpret_cast<SLuint32*>(value->data));
+            SLuint32 channels;
+            memcpy(&channels, value->data, sizeof(SLuint32));
             LOGD("channels: %d", channels);
             *channelsOut = channels;
           }
diff --git a/variablespeed/jni/variablespeed.h b/variablespeed/jni/variablespeed.h
index 07cba0f..cf856da 100644
--- a/variablespeed/jni/variablespeed.h
+++ b/variablespeed/jni/variablespeed.h
@@ -86,7 +86,7 @@ class AudioEngine {
   bool HasSampleRateAndChannels();
   SLuint32 GetSLSampleRate();
   SLuint32 GetSLChannels();
-  size_t GetChannelCount();
+  SLuint32 GetChannelCount();
 
   // The single global audio engine instance.
   static AudioEngine* audioEngine_;
-- 
1.8.1.6

